conbench/entities/history.py

```diff
@@ -255,58 +254,63 @@ def get_history_for_cchr(
-    # For most invocations of this function, there are very few unique run_ids among the
-    # benchmark_results. Sort them by run_id and run aoptimized query for each group.
-    sorted_by_run_id: DefaultDict[str, List[BenchmarkResult]] = defaultdict(list)
+    run_ids = set(result.run_id for result in benchmark_results)
+    if len(run_ids) != 1:
+        raise ValueError(f"Encountered mixed run_ids in set_z_scores(): {run_ids}")
+    run_id = run_ids.pop()
+
+    if len(benchmark_results) == 1:
+        # performance optimization
+        case_id = benchmark_results[0].case_id
+        context_id = benchmark_results[0].context_id
+    else:
+        case_id = None
+        context_id = None
+
+    distribution_stats = _query_distribution_stats_by_run_id(
+        run_id=run_id,
+        baseline_commit=baseline_commit,
+        case_id=case_id,
+        context_id=context_id,
+    )
+
     for benchmark_result in benchmark_results:
-        sorted_by_run_id[benchmark_result.run_id].append(benchmark_result)
-
-    for run_id, result_group in sorted_by_run_id.items():
-        if len(result_group) == 1:
-            # performance optimization
-            case_id = result_group[0].case_id
-            context_id = result_group[0].context_id
-        else:
-            case_id = None
-            context_id = None
-
-        distribution_stats = _query_distribution_stats_by_run_id(
-            run_id, case_id=case_id, context_id=context_id
+        dist_mean, dist_stddev = distribution_stats.get(
+            (benchmark_result.case_id, benchmark_result.context_id), (None, None)
+        )
+        benchmark_result.z_score = _calculate_z_score(
+            data_point=_to_float(benchmark_result.mean),
+            unit=benchmark_result.unit,
+            dist_mean=_to_float(dist_mean),
+            dist_stddev=_to_float(dist_stddev),
         )
-
-        for benchmark_result in result_group:
-            dist_mean, dist_stddev = distribution_stats.get(
-                (benchmark_result.case_id, benchmark_result.context_id), (None, None)
-            )
-            benchmark_result.z_score = _calculate_z_score(
-                data_point=_to_float(benchmark_result.mean),
-                unit=benchmark_result.unit,
-                dist_mean=_to_float(dist_mean),
-                dist_stddev=_to_float(dist_stddev),
-            )
 
 
 def _query_distribution_stats_by_run_id(
-    run_id: str, case_id: Optional[str], context_id: Optional[str]
+    run_id: str,
+    baseline_commit: Commit,
+    case_id: Optional[str],
+    context_id: Optional[str],
 ) -> Dict[Tuple[str, str], Tuple[Optional[float], Optional[float]]]:
-    """Given a run_id, return stats of the distribution of BenchmarkResults in the run's
-    commit ancestry, by case/context. Returns a dict that looks like
+    """Given a run id and baseline commit, return stats of the distribution of
+    BenchmarkResults matching the run's hardware, in the baseline_commit's ancestry, by
```

Comment:
> return stats of the distribution of  BenchmarkResults matching the run's hardware

Of _all_ `BenchmarkResults`, or just those that are associated with that Run?

Reply:
I'll clarify this in an upcoming PR (see my above comment)